internal/testing/cmdtest/cmdtest.go

```diff
@@ -0,0 +1,610 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// The cmdtest package simplifies testing of command-line interfaces. It
+// provides a simple, cross-platform, shell-like language to express command
+// execution. It can compare actual output with the expected output in the file,
+// and can also update a file with new "golden" output that is deemed correct.
+//
+// Start using cmdtest by writing a test file with commands and expected output.
+// See the TestFile documentation for the syntax.
+//
+// To test, first read the file:
+//
+//    tf, err := cmdtest.Read("test.ct")
+//
+// Then configure the resulting TestFile by adding commands or enabling
+// debugging features. Lastly, call TestFile.Compare:
+//
+//    diff := tf.Compare()
+//
+// The string returned by Compare will be non-empty if there were any
+// errors or differences in output.
+//
+// After confirming that the output differences are intentional, you can
+// update the original file by calling TestFile.Update:
+//
+//    err := tf.Update()
+package cmdtest
+
+import (
+	"bufio"
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+// A TestFile describes a single file that may contain multiple test cases.
+// File format:
+//
+// Before the first line starting with a '$', empty lines and lines beginning with
+// "#" are ignored.
+//
+// A sequence of consecutive lines starting with '$' begin a test case. These lines
+// are commands to execute. See below for the valid commands.
+//
+// Lines following the '$' lines are command output (merged stdout and stderr).
+// Output is always treated literally. After the command output there should be a
+// blank line. Between that blank line and the next '$' line, empty lines and lines
+// beginning with '#' are ignored. (Because of these rules, cmdtest cannot
+// distinguish trailing blank lines in the output.)
+//
+// Syntax of a line beginning with '$':
+// A sequence of space-separated words (no quoting is supported). The first word is
+// the command, the rest are its args. If the next-to-last word is '<', the last word
+// is interpreted as a file and becomes the standard input to the command. This input
+// redirection is only supported for commands run with exec.Command, not those in the
+// Commands map.
+//
+// By default, commands are expected to succeed, and the test will fail
+// otherwise. However, commands that are expected to fail can be marked
+// with a " --> FAIL" suffix.
+//
+// The cases of a test file are executed in order, starting in a freshly
+// created temporary directory.
+//
+// The built-in commands (initial contents of the Commands map) are:
+//
+//   cd DIR
+//   cat FILE
+//   mkdir DIR
+//   setenv VAR VALUE
+//   echo ARG1 ARG2 ...
```

Comment:
Why not add `ls`? It seems like it might be common, and it somewhat subtle due to platform-specific restrictions.

Reply:
Omitting it for now, as per our discussion.